File: dnn-providers/hip-kernel-provider/rocke/platform/cpp/instances/common/_moe_fused_mega_lds.cpp
// Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
// SPDX-License-Identifier: MIT
/*
 * _moe_fused_mega_lds.cpp -- C99 port of
 * rocke/instances/common/_moe_fused_mega_lds.py: the whole-kernel LDS
 * accounting shared by the fused-MoE mega-kernel family.
 *
 * See rocke/helper_rocke.instances.common._moe_fused_mega_lds.h for the symbol
 * map and for why summing the aligned segments is EXACT (not conservative) for
 * the megas. None of these routines emit IR; each mega's build entry calls
 * rocke_validate_mega_lds_budget with the alloc sequence its own _lds_allocs
 * produces, in emitter declaration order.
 */
#include "rocke/helper_rocke.instances.common._moe_fused_mega_lds.h"

#include <stdio.h>
#include <string.h>

#include "rocke/arena.h" /* rocke_arena_printf (reason strings)                */
#include "rocke/helper_rocke.core.arch.h" /* rocke_archtarget_* (ArchTarget)   */
#include "rocke/helper_rocke.helpers.mfma_gemm_inner.h" /* validate_arch_and_block_size */

/* Longest reject breakdown the shipped megas produce is ~110 chars (five/six
 * buffers); the assembled message is then bounded again by ROCKE_ERR_MSG_CAP at
 * the caller's rocke_i_set_err. Truncation here would only shorten a
 * human-readable reject reason -- it never reaches IR. */
#define ROCKE_MEGA_LDS_BREAKDOWN_CAP 512

long rocke_mega_lds_alloc_nbytes(const rocke_mega_lds_alloc_t* alloc)
{
    if(alloc == NULL)
    {
        return 0;
    }
    return (long)alloc->elem_bytes * (long)alloc->elem_count;
}

int rocke_mega_lds_elem_bytes(const rocke_type_t* dtype)
{
    /* Mirror of lower_llvm's _elem_bytes map (and its width-2 fallback) so the
     * accounting cannot disagree with the packer. */
    const char* n = (dtype != NULL) ? dtype->name : NULL;
    if(n == NULL)
    {
        return 2;
    }
    if(strcmp(n, "i8") == 0 || strcmp(n, "fp8e4m3") == 0 || strcmp(n, "bf8e5m2") == 0)
    {
        return 1;
    }
    if(strcmp(n, "f16") == 0 || strcmp(n, "bf16") == 0)
    {
        return 2;
    }
    if(strcmp(n, "i32") == 0 || strcmp(n, "f32") == 0)
    {
        return 4;
    }
    if(strcmp(n, "i64") == 0)
    {
        return 8;
    }
    return 2;
}

/* _seg_align(elem_bytes): the segment alignment the packer applies -- 16 B for
 * byte-element types, else 4. */
static long rocke_mega_lds_seg_align(int elem_bytes)
{
    return (elem_bytes == 1) ? 16 : 4;
}

long rocke_mega_lds_pool_bytes(const rocke_mega_lds_alloc_t* allocs, size_t num_allocs)
{
    long end = 0;
    size_t i;

    if(allocs == NULL)
    {
        return 0;
    }
    for(i = 0; i < num_allocs; ++i)
    {
        long aln = rocke_mega_lds_seg_align(allocs[i].elem_bytes);
        end = ((end + aln - 1) & ~(aln - 1)) + rocke_mega_lds_alloc_nbytes(&allocs[i]);
    }
    return (end + 15) & ~15L;
}

bool rocke_validate_mega_lds_budget(rocke_ir_builder_t* b,
                                    const rocke_mega_lds_alloc_t* allocs,
                                    size_t num_allocs,
                                    const char* arch,
                                    const char** out_reason)
{
    const rocke_archtarget_t* target = NULL;
    const char* arch_reason = NULL;
    char breakdown[ROCKE_MEGA_LDS_BREAKDOWN_CAP];
    size_t pos = 0;
    size_t i;
    long total;

    /* try: target = ArchTarget.from_gfx(arch) / except KeyError as e: return
     * False, str(e).
     *
     * block_size 0 can never exceed a target's max_threads_per_block, so this
     * call reduces to exactly that from_gfx lookup plus the one reconstruction
     * of the Python KeyError repr the engine owns. (Unreachable from either
     * mega in practice: both run an arch gate before this one.) */
    if(!rocke_validate_arch_and_block_size(b, arch, 0, &arch_reason, &target) || target == NULL)
    {
        if(out_reason != NULL)
        {
            *out_reason = (arch_reason != NULL) ? arch_reason : "unknown gfx target";
        }
        return false;
    }

    total = rocke_mega_lds_pool_bytes(allocs, num_allocs);
    if(rocke_archtarget_fits_lds(target, total))
    {
        if(out_reason != NULL)
        {
            *out_reason = "ok";
        }
        return true;
    }

    /* breakdown = ", ".join(f"{a.name}={a.nbytes}" for a in allocs) */
    breakdown[0] = '\0';
    for(i = 0; i < num_allocs && allocs != NULL && pos + 1 < sizeof(breakdown); ++i)
    {
        int wrote = snprintf(breakdown + pos,
                             sizeof(breakdown) - pos,
                             "%s%s=%ld",
                             (i == 0) ? "" : ", ",
                             (allocs[i].name != NULL) ? allocs[i].name : "None",
                             rocke_mega_lds_alloc_nbytes(&allocs[i]));
        if(wrote < 0)
        {
            break;
        }
        pos += (size_t)wrote;
        if(pos >= sizeof(breakdown))
        {
            pos = sizeof(breakdown) - 1; /* snprintf truncated; stop appending */
            break;
        }
    }

    if(out_reason != NULL)
    {
        *out_reason = (b != NULL) ? rocke_arena_printf(&b->arena,
                                                       "LDS budget %ld > %d cap (%s) on %s",
                                                       total,
                                                       target->lds_capacity_bytes,
                                                       breakdown,
                                                       (arch != NULL) ? arch : "None")
                                  : "LDS budget over cap";
    }
    return false;
}

File: dnn-providers/hip-kernel-provider/rocke/platform/python/rocke/instances/common/_moe_fused_mega_lds.py
# Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
# SPDX-License-Identifier: MIT

"""Whole-kernel LDS accounting shared by the fused-MoE mega-kernel family.

Each mega (``moe_fused_mega`` f16/bf16, ``moe_fused_mega_fp8``, and the gfx1250
``fused_moe_mega_wmma``) validates its gate/up and its down GEMM as two
INDEPENDENT :class:`UniversalGemmSpec` s. Neither sub-validation sees
``Hidden_smem`` -- the persistent LDS bridge that is the whole point of the
fusion -- nor the fact that both GEMMs' operand buffers are allocated in the
builder prologue and therefore all coexist. A mega whose two halves each fit the
per-WG budget can still blow it as a whole; without this module that only
surfaces as a kernel-load failure on the device instead of a spec rejection.

Why a plain sum is EXACT here, not conservative
-----------------------------------------------
``core/lower_llvm.py`` packs every ``smem_alloc`` into one ``@smem_pool`` global
with a liveness-driven linear scan, so an allocation *may* be placed on top of a
dead one -- the cshuffle epilogue in ``gemm_universal`` relies on exactly that.
It cannot happen for the megas: a live interval opens at the ``tile.smem_alloc``
op, and every mega buffer is allocated in the prologue, before any of them is
first used. All pairs therefore interfere by construction and the packer gives
each a disjoint range, so the down GEMM's ``Bd_smem`` does **not** alias the
by-then-dead ``Bg_smem`` / ``Bu_smem``. Summing the aligned segments reproduces
``_smem_pool_size`` byte for byte; each mega's ``tests/instances/
test_moe_fused_mega*.py`` asserts that equality against the lowered IR so the
accounting and the packer cannot drift.

Allocations the emitter declares but never references are dead-stripped by the
packer and must be left out of the sequence passed here (the fp8 mega's
``BStage_smem`` under ``use_dtla=False`` is the live example).
"""

from __future__ import annotations

from typing import NamedTuple, Sequence, Tuple


__all__ = [
    "LdsAlloc",
    "lds_elem_bytes",
    "mega_lds_pool_bytes",
    "validate_mega_lds_budget",
]


# Mirror of ``lower_llvm._Lowerer._compute_smem_layout``'s ``_elem_bytes`` map
# (and its width-2 fallback) so the accounting cannot disagree with the packer.
_ELEM_BYTES = {
    "i8": 1,
    "fp8e4m3": 1,
    "bf8e5m2": 1,
    "f16": 2,
    "bf16": 2,
    "i32": 4,
    "f32": 4,
    "i64": 8,
}


class LdsAlloc(NamedTuple):
    """One ``smem_alloc`` as the LDS budget sees it.

    ``name`` is the emitter's ``name_hint`` (it names the buffer in the
    rejection message); ``elem_bytes`` and ``elem_count`` give the segment size.
    """

    name: str
    elem_bytes: int
    elem_count: int

    @property
    def nbytes(self) -> int:
        return self.elem_bytes * self.elem_count


def lds_elem_bytes(dtype) -> int:
    """Bytes per element of an IR scalar type, as the smem packer measures it."""
    return _ELEM_BYTES.get(getattr(dtype, "name", dtype), 2)


def _seg_align(elem_bytes: int) -> int:
    """Segment alignment the packer applies: 16 B for byte-element types, else 4."""
    return 16 if elem_bytes == 1 else 4


def mega_lds_pool_bytes(allocs: Sequence[LdsAlloc]) -> int:
    """Bytes the ``@smem_pool`` global occupies for ``allocs``.

    Replays the packer's placement for the all-interfering mega case: each
    segment starts at the next multiple of its alignment past the previous one,
    and the pool is rounded up to 16 B. ``allocs`` must be in emitter
    declaration order.
    """
    end = 0
    for alloc in allocs:
        aln = _seg_align(alloc.elem_bytes)
        end = ((end + aln - 1) & ~(aln - 1)) + alloc.nbytes
    return (end + 15) & ~15


def validate_mega_lds_budget(allocs: Sequence[LdsAlloc], arch: str) -> Tuple[bool, str]:
    """Return ``(ok, reason)`` for a mega's total LDS against ``arch``'s budget."""
    from ...core.arch import ArchTarget

    try:
        target = ArchTarget.from_gfx(arch)
    except KeyError as e:
        return False, str(e)
    total = mega_lds_pool_bytes(allocs)
    if target.fits_lds(total):
        return True, "ok"
    breakdown = ", ".join(f"{a.name}={a.nbytes}" for a in allocs)
    return False, (
        f"LDS budget {total} > {target.lds_capacity_bytes} cap "
        f"({breakdown}) on {arch}"
    )
